overlord/fdestate/extop.go

```diff
@@ -0,0 +1,162 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+package fdestate
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/snapcore/snapd/overlord/state"
+)
+
+// ExternalOperationStatus captures the status of operations running externally.
+type ExternalOperationStatus int
+
+// Admitted status values for changes and tasks.
+const (
+	DefaultStatus ExternalOperationStatus = 0
+
+	// Preparing means that we are performing preparation steps, but the
+	// operation isn't yet running externally.
+	PreparingStatus ExternalOperationStatus = 1
+
+	// DoingStatus means the operation is running externally.
+	DoingStatus ExternalOperationStatus = 2
+
+	// DoneStatus means the operation has completed successfully.
+	DoneStatus ExternalOperationStatus = 3
+
+	// CompletingStatus means the operation is completing.
+	CompletingStatus ExternalOperationStatus = 4
+
+	// AbortingStatus means the operation is aborting.
+	AbortingStatus ExternalOperationStatus = 5
+
+	// ErrorStatus means the operation has failed.
+	ErrorStatus ExternalOperationStatus = 6
+)
+
+func (s ExternalOperationStatus) String() string {
+	switch s {
+	case DefaultStatus:
+		return "default"
+	case PreparingStatus:
+		return "preparing"
+	case DoingStatus:
+		return "doing"
+	case DoneStatus:
+		return "done"
+	case CompletingStatus:
+		return "completing"
+	case AbortingStatus:
+		return "aborting"
+	case ErrorStatus:
+		return "error"
+	}
+	panic(fmt.Sprintf("internal error: unexpected external operation status code: %d", s))
+}
+
+type externalOperation struct {
+	Kind string `json:"kind"`
+	// ChangeID is ID of a state.Change associated with the operation.
+	ChangeID string                  `json:"change-id"`
+	Status   ExternalOperationStatus `json:"status"`
+	// Err when not empty, carries the error message, usually associated with
+	// the operation having an ErrorStatus or AbortingStatus.
+	Err string `json:"err"`
+	// Context is an opaque piece of data associated with the operation.
+	Context json.RawMessage `json:"context"`
+}
+
+func (e *externalOperation) SetFailed(msg string) {
+	e.Status = ErrorStatus
+	e.Err = msg
+}
+
+func (e *externalOperation) SetStatus(status ExternalOperationStatus) {
```

Comment:
same for this

Reply:
this is still not tested

Reply:
pushed a tweak in https://github.com/canonical/snapd/pull/14764/commits/f99ed68e832ac11b0793d5b434335bce686106c7